proposals/text/5540-llm-support-2.0.md

```diff
@@ -0,0 +1,309 @@
+
+In Haystack 1.x, `PromptNode` uses `InvocationLayer` to query different LLMs under a unified API. In that design, users do not need to know which invocation layer is used for the model they select, as `PromptNode` takes responsibility of selecting it.
+
+Such invocation layers can be ported to 2.0 as standalone components. In this way we will have one component for each LLM backed that we support.
+
+Each component should be named after the class of models it supports. For example we should have `GPT4`, `ChatGPT`, etc. For components whose name may be confusing, or that include models that are not LLMs, we can add a LLM suffix: like `FalconLLM`, `CohereLLM`, `HuggingFaceLLM`.
+
+Note that having separate components for each LLM makes easy to deprecate them when we realize they are dropping out of favor or become severely outdated. It also makes very easy for external contributors to make their own external components to support rarer LLMs, without having to add them to Haystack’s core.
+
+All these LLM clients will have a near-identical I/O:
+
+```python
+@component
+class ChatGPT:
+
+    @component.output_types(replies=List[List[str]])
+    def run(self, prompts: List[str], ... chatgpt specific params...):
+        ...
+        return {'replies': [...]}
+```
+
+Note how the component takes a list of prompts and LLM parameters only, but no variables nor templates, and returns only strings. This is because input rendering and output parsing are delegated to separate components, which description follows.
+
+### Returning metadata
+
+In the example above we made the LLM return only a list of replies, as strings. However, in order to be able to parse the output into meaningful objects (see “Output parsing”) we may need additional metadata from these clients.
+
+1. Do we already have any such situation?
+2. Can we foresee any other?
+
+If the answer to any of the above is yes, a simple, maybe temporary solution would be to add a second output, called for example `'replies_meta'` . Any component that need such meta to parse the output would then request this second output along with the first and zip the two lists together to reconstruct the original output of the LLM.
+
+### Returning streams
+
+In some cases users may want to see the output of the LLM as it’s being generated, so as a stream that prints word by word in the terminal.
+
+To achieve such effect, we should make LLM components return a more generic type `replies: List[Iterator[str]]`. In this way, users will have the ability of returning a generator or iterator in place of a regular string, and therefore allow the receiving components to unroll these replies as they wish. Components that need the full reply will unroll the iterator internally (like `RepliesToAnswersConverter`), while if the generator goes in the output, the users can unroll it in a loop printing the content word by word. We can also have specialized components to visualize the stream within a pipeline: something like:
```

Comment:
How is streaming handled within the pipeline? Can you hook up a pipeline to a REST API and then stream the pipeline output?

Reply:
This is what the paragraph describes, right? If we make LLMs return an iterator instead of a fully in-memory list of tokens, that should behave a bit like a promise: it's evaluated lazily, so it's returned immediately and the user is able to "unroll" it once they receive it as output.

So if the LLM outputs directly to the user (let's imagine the "user" is the REST API) I can convert that into a stream in the endpoint and stream it back to the client. While if the LLM outputs to a component that needs the full output before moving on, we can put a "stream consumer" in between them, that unrolls the iterator and then passes the results to the following component once all the data have arrived. It might print to console in the meantime to give users some feedback as well.

Is this the feature you're looking for? Or is it something different?

Reply:
This is something that concerns me a lot and I think we're skimming over it waaaay too lightly. 
It's a topic that could entail its own proposal in my opinion.

Returning an `Iterator` is not the correct way to go I think for several reasons. 
First of all `Iterator` doesn't necesseraly means `Generator`, a `Generator` respects the `Iterator` type but not viceversa. That doesn't mean it's enough to just change the type to `Generator`.

Another issue that I see is that this might not actually be what users want. From my understanding people want to consume streamed HTTP responses as they are received.

Returning a generator might seem like a good solution but it opens up quite some pitfalls.

Say that the generator is passed as input to two other components and they both try to consume it at the same time. With the current synchronous Pipeline execution one component gets the whole generator output and the other none. If and when we add parallel branch execution that will only get worse as there could be multiple components consuming the same generator at the same time getting different chunks.

Should we wait for the generator to be consumed to make the first HTTP request? Shall we instead make the first request in the prompt component that creates the generator? Why even return an iterator at this point though? The backend service will keep sending tokens even if the generator is not being actively consumed. This is bad in my opinion, we're telling the users that the response is being streamed as it's received but that's not true. 
That might make the users think that stopping consumption severes the connection and the token generation on the other side, but that's not true.

All in all I still have no clear solution for this but I don't think we should tackle this in this proposal. The idea present in here is similar to the one we have in v1 right now and that's bit added quite hastily, I would prefer if we think it more thoroughly before digging.

It's a huge topic that requires quite some discussion to plan it out correctly and I'd leave it out of this proposal as of now.

Reply:
@ZanSara so this would work if the prompt node is the last node in a pipeline? What about parsers, would you just set the answer content to the generator and then return from the pipeline?

Agree with @silvanocerza that this is a very complex topic.

Reply:
Ok you bring out some very good points here... and that's true we don't have any solution for it at the moment.

Do you believe we should still consider having a "patch" with the iterators? Or is it better to "block" this usecase until a later time when we can get it done properly? I'm not under the impression that this might have to do with the discussion on async pipelines (which should not bleed into this proposal, as it's a huge topic on its own). That is one of the next topic in line for proposals anyway, so postponing it to that time doesn't mean forgetting about it at all.

Reply:
The streaming support is a separate topic from the async support for me 👍  I'm happy with discussing streaming in another proposal / place.

Reply:
Ok! I'll collect ideas on this topic separately then :+1:

Reply:
Collected here: https://www.notion.so/deepsetai/Notes-about-streaming-the-LLM-outputs-7ca49b892237449fad59bd0e7cc03e69 I'm resolving this to to clear up the page, not closing the topic at all